companion/src/eepromimportexport.h

```diff
@@ -698,8 +698,11 @@ class ConversionField: public TransformedField {
       if (table) {
         if (table->exportValue(_field, _field))
           return;
-        if (!error.isEmpty())
+        if (!error.isEmpty()) {
+          // TODO how could we avoid this global?
```

Comment:
What is the ultimate goal in this particular case?  I only see `EEPROMWarnings` displayed from inside `OpenTxEeepromInterface::save()`.

The following idea does a bit of assuming but I'll toss it out there anyway.

If each field is a subtype of `DataField` then it would make a lot of sense to have an error/warning handler in the parent class (or perhaps in `TransformedField` if `DataField` is too broad).  I'm not sure why you suggest the parent/child relationship, but if you need to track a hierarchy then that could be built into the handler (you could pass the parent and child object names for example, along with the message and e.g. error code). Of course the parent class already has some of that info, like `name`, so that could easily be added to the warning data already.  E.g. just something as simple as `DataField::addWarning(const QString &txt)` would do it, no?

As to how to actually get the error "collector" in there (`EEPROMWarnings`), I can think of a few ways... but here's one.  Since `EEPROMInterface` already essentially exists as a singleton, I'd add the "collector" as a member of that (eg. it could be a public struct definition with more data than just message).  Then in the `DataField` (or `TransformedField`) initializer I'd get a reference to the current `EEPROMInterface` instance (`EEPROMInterface * eeInstance = getCurrentEEpromInterface();`).  I'd then use whatever function(s) `EEPROMInterface` provides for adding warning messages (I'd suggest a function rather than direct access to the "collector" variable).  Or in other words inside our theoretical `DataField::addWarning(txt)` function we'd just call something like `eeInstance->addWarning(txt);`.

Actually I'd probably make `EEPROMInterface` and `DataField` inherit from QObject and then connect `DataField::warningAdded()` to `EEPROMInterface::onWarningAdded()` (or something like that).

Of course `EEPROMWarnings` could also be its own self-contained independent singleton/static class/struct, but I'm not sure it would add much benefit.  Or a part of new class for that matter, eg. `EepromImportExport`.  At that point it becomes a question of if that same warnings container would be useful for other aspects of dealing with `EEPROMInterface`, or if it's purely for import/export.